velox/exec/tests/MultiFragmentTest.cpp

```diff
@@ -787,3 +788,147 @@ TEST_F(MultiFragmentTest, earlyCompletionMerge) {
     ASSERT_TRUE(waitForTaskCompletion(task.get())) << task->taskId();
   }
 }
+
+class CustomSlowNode : public core::PlanNode {
+ public:
+  CustomSlowNode(const core::PlanNodeId& id, core::PlanNodePtr source)
+      : PlanNode(id), sources_{std::move(source)} {}
+
+  const RowTypePtr& outputType() const override {
+    return sources_[0]->outputType();
+  }
+
+  const std::vector<core::PlanNodePtr>& sources() const override {
+    return sources_;
+  }
+
+  std::string_view name() const override {
+    return "custom slow";
+  }
+
+ private:
+  void addDetails(std::stringstream& /* stream */) const override {}
+
+  std::vector<core::PlanNodePtr> sources_;
+};
+
+class CustomSlowOperator : public Operator {
+ public:
+  CustomSlowOperator(
+      int32_t operatorId,
+      DriverCtx* driverCtx,
+      std::shared_ptr<const CustomSlowNode> slowNode)
+      : Operator(
+            driverCtx,
+            slowNode->outputType(),
+            operatorId,
+            slowNode->id(),
+            "CustomSlowOperator") {}
+
+  void addInput(RowVectorPtr input) override {
+    inputs_.emplace_back(std::move(input));
+  }
+
+  bool needsInput() const override {
+    return inputs_.empty();
+  }
+
+  RowVectorPtr getOutput() override {
+    if (inputs_.empty() || consumedOneBatch) {
+      return nullptr;
+    }
+    auto result = inputs_.back();
+    inputs_.pop_back();
+    consumedOneBatch = true;
+    return result;
+  }
+
+  void noMoreInput() override {
+    Operator::noMoreInput();
+  }
+
+  BlockingReason isBlocked(ContinueFuture* future) override {
+    return BlockingReason::kNotBlocked;
+  }
+
+  bool isFinished() override {
+    return false;
+  }
+
+ private:
+  std::vector<RowVectorPtr> inputs_;
+  bool consumedOneBatch{false};
+};
+
+class CustomSlowOperatorTranslator : public Operator::PlanNodeTranslator {
+  std::unique_ptr<Operator>
+  toOperator(DriverCtx* ctx, int32_t id, const core::PlanNodePtr& node) {
+    if (auto slowNode = std::dynamic_pointer_cast<const CustomSlowNode>(node)) {
+      return std::make_unique<CustomSlowOperator>(id, ctx, slowNode);
+    }
+    return nullptr;
+  }
+};
+
+TEST_F(MultiFragmentTest, exchangeDestruction) {
+  // This unit test tests the proper destruction of ExchangeClient upon
+  // task destruction.
+  OperatorTestBase::SetUp();
+  Operator::registerOperator(std::make_unique<CustomSlowOperatorTranslator>());
+  ExchangeSource::factories().clear();
+  ExchangeSource::registerFactory([](const std::string& taskId,
+                                     int destination,
+                                     std::shared_ptr<ExchangeQueue> queue,
+                                     memory::MemoryPool* pool) {
+    if (strncmp(taskId.c_str(), "local://", 8) == 0) {
+      return std::make_unique<LocalExchangeSource>(
```

Comment:
This factory looks very similar to default one. What's the difference?

Reply:
This factory passes a pool but the default one pass nullptr to the pool param

Reply:
Is there any reason for the default factor to pass null. Can we update the default to pass the pool instead (in the PR doing the refactoring)?

Reply:
@tanjialiang Is it necessary to use a cluster Exchange Source? Looks like the default can work just fine, no?